py/src/nupic/optimization/ae.py

```diff
@@ -124,6 +128,14 @@
 
 from nupic.optimization.parameter_set import ParameterSet
 
+acceptable_exceptions = [
+    TypeError,
+    ValueError,
+    MemoryError,
+    ZeroDivisionError,
+    AssertionError,
+    RuntimeError,]
```

Comment:
The optimization program has no knowledge of the valid ranges of the parameters.  It tries parameters which will fail.  For example it tried a negative sparsity, which set off an `assert` statement.

Also, these are not totally ignored.  The resulting stack traces are saved to the experiment's journal file, and the hash of the parameters which caused the exception is listed under "Failed Experiments".

By "Ignored" what I mean is that the optimization program should continue running.

Reply:
> The optimization program has no knowledge of the valid ranges of the parameters. It tries parameters which will fail. For example it tried a negative sparsity, which set off an assert statement.

not in this PR, but would be nice if the experiment framework checked for the valid ranges of params (ie >= 0, or in range [0.0, 1.0]. Is it possible to specify these, and the step/diff size? 

On the other hand, this is sort of an input fuzzying, so it checks our methods have input-checking in place.

Reply:
While yes, it would be nice if the optimization program was aware of the constraints on the parameters, it's a lot of work for the user to specify them, and all of our methods already have input-checking.  As a user, I'd rather raise an exception after being given bad parameters than have to type out the valid ranges of the parameters and communicate them to the optimizer program.

Reply:
>  As a user, I'd rather raise an exception after being given bad parameters than have to type out the valid ranges of the parameters and communicate them to the optimizer program. 

say a parameter takes only values in range `0.01..0.02`, how can be be sure without the ranges & step size that the optimizer tries any (number of) meaningful combinations at all?

Reply:
We can't be sure, other than by checking the lab-report file and searching for a modification to that parameter which did not fail.

However, most of the optimizer try to modify the parameters by a small percent.  For example grid search tries to modify the value `v` by the following percentages:
```
v *  .40
v *  .60
v *  .80
v * 1.20
v * 1.40
v * 1.60
```
So your example where the valid range is `[.01, .02]` would work because the initial value (given by the user in their `default_parameters`) would be in that range too, and a 20% or 40% change in either direction could remain inside of the valid range.

Reply:
can this be a bit automated, controlled? 

> We can't be sure, other than by checking the lab-report file and searching for a modification to that parameter which did not fail. 

say have a param least_n_tries (=5) and each (or make this param specific) param must be tried at least that times? 

>  grid search tries to modify the value v by the following percentages:
v *  .40
v *  .60

this is basically the step size, can that be a default param overridable by the users' liking? 

> would work because the initial value (given by the user in their default_parameters) would be in that range too

check that the given default param combination is valid, and fail on the user if not? (bcs if given totally out of plave, we cannot be sure any useful params are found and tested?) 

> and a 20% or 40% change in either direction could remain inside of the valid range.

and also check that at least some of the diffs/steps are also valid? 

That way, user does not have to manually specify the ranges and steps (although I think it's not too much work and the serach-space would be much smaller if hand tailored, thus reaching better results during the same time for optimization) , 
and we'd still ensure atleast some combinations are tried out and valid.

Reply:
> say have a param least_n_tries (=5) and each (or make this param specific) param must be tried at least that times?

Grid search tries each modification in an infinite loop, so if you wait long enough it will try each parameter (and each modification of it) 5 times.

> can that be a default param overridable by the users' liking?

There is a TODO note in the source code to do this, but for me its a low priority.

> check that the given default param combination is valid

Yes, we can add a special case for when the default_parameters fail.  If the default parameters don't work then the user needs to fix them before running the optimization programs.